bpo-34037: Fix test_asyncio failure and add loop.shutdown_default_executor()

aeros opened the pull request:
This PR was based heavily upon @vstinner's [PR](https://github.com/python/cpython/pull/13786) and the suggestions from @asvetlov to improve upon it:

>  I support the idea but have a question about implementation.
asyncio provides an interface class for loop called AbstractEventLoop.
I believe that the new property should be reflected there.
The next question is: should it be a property or get_wait_executor_on_close() / set_wait_executor_on_close() pair of methods? Asyncio loop is designed to don't use properties but getters / setters. 

@vstinner's PR was reverted due to it making too drastic of an API change for 3.8 and the concerns from @asvetlov about adding the property without providing public getter and setter methods.

However, the changes proposed still had significant support, and **most importantly**, it fixes a bug which has been causing intermittent failures within the CI tests for ``test_asyncio`` due to dangling threads:
```
test_run_in_executor_cancel (test.test_asyncio.test_events.KqueueEventLoopTests) ...

  Warning -- threading_cleanup() failed to cleanup 1 threads (count: 1, dangling: 2)
  Dangling thread: <Thread(ThreadPoolExecutor-13_0, started daemon 34471091200)>
  Dangling thread: <_MainThread(MainThread, started 34393318400)>
```

Since this is involving changes to the public API, I fully plan on adding the appropriate changes to the documentation. But, I would prefer to wait until I receive feedback from @vstinner and @asvetlov about the implementation itself before doing so since they were both significantly involved in @vstinner's previous PR.

I'll cc the asyncio team as well in case anyone else has feedback to provide.

/cc @python/asyncio 


<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-34037](https://bugs.python.org/issue34037) -->
https://bugs.python.org/issue34037
<!-- /issue-number -->

aeros commented on Misc/NEWS.d/next/Library/2019-09-09-04-43-26.bpo-34037.kX3ZqZ.rst:
```diff
@@ -0,0 +1 @@
+For :mod:`asyncio`, add attribute ``asyncio.BaseEventLoop.wait_executor_on_close`` to provide an option for waiting for the executor before closing the event loop. Also adds methods ``loop.get_wait_executor_on_close`` and ``loop.set_wait_executor_on_close``.
```
I'll fix the number of characters per line and add the appropriate Sphinx roles after I finish adding the documentation changes.

vstinner commented on Lib/asyncio/base_events.py:
```diff
@@ -1867,3 +1870,13 @@ def set_debug(self, enabled):
+    def set_wait_executor_on_close(self, enabled):
+        if not isinstance(enabled, bool):
+            raise TypeError(f'Expected enabled to be boolean,'
+                            f'got {enabled!r}')
```
please add a space before "got"

vstinner commented on Misc/NEWS.d/next/Library/2019-09-09-04-43-26.bpo-34037.kX3ZqZ.rst:
```diff
@@ -0,0 +1 @@
+For :mod:`asyncio`, add attribute ``asyncio.BaseEventLoop.wait_executor_on_close`` to provide an option for waiting for the executor before closing the event loop. Also adds methods ``loop.get_wait_executor_on_close`` and ``loop.set_wait_executor_on_close``.
```
it's not an attribute but 2 new methods.

aeros replied:
@vstinner:
> it's not an attribute but 2 new methods.

When I initially created the news entry, the ``wait_executor_on_close`` attribute/property (within Python, I believe property would probably be the more correct term) was public, as it was in your PR. But then I realized that it should probably be private since the getter and setter methods were created. 

I'll update it accordingly.

vstinner reviewed (dismissed):
You have to update the documentation:

* Document the 2 new event loop methods, around https://docs.python.org/dev/library/asyncio-eventloop.html
* Document the behavior change in What's New in Python 3.9, around: https://docs.python.org/dev/whatsnew/3.9.html#changes-in-the-python-api

1st1 reviewed (changes requested):
I still don't like this API and I'd suggest to implement it the other way around. We have a very similar problem with asynchronous generators -- closing them isn't an easy process and can be a lengthy blocking operation.  The way we do that is via the `loop.shutdown_asyncgens()` method, that users are supposed to call *manually* before `loop.close()`.  This is a low-level API, and the expectation is that `asyncio.run()` function would do that for you.

Therefore I propose the following:

* Add a new `loop.shutdown_threadpool()` method.  Just like with `shutdown_asyncgens()` -- it would be invalid to call `loop.run_in_executer()` after `loop.shutdown_threadpool()` is called.

* Make `asyncio.run()` to call the new `loop.shutdown_threadpool()`.

aeros committed:
Add flag BaseEventLoop._executor_shutdown_called

aeros committed:
Add abstract method for loop.shutdown_default_executor()

aeros committed:
Add BaseEventLoop method shutdown_default_executor()

aeros committed:
Remove whitespace

aeros committed:
Add BaseEventLoop internal method _check_executor_shutdown()

aeros committed:
Adjust docstring for abstract method loop.shutdown_default_executor

aeros committed:
Rename method _check_executor_shutdown() to _check_default_executor()

aeros committed:
Add default executor check to loop.run_in_executor()

aeros committed:
Implement loop.shutdown_default_executors() in asyncio.run()

1st1 commented on Lib/asyncio/base_events.py:
```diff
@@ -404,8 +404,11 @@ def __init__(self):
         # A weak set of all asynchronous generators that are
         # being iterated by the loop.
         self._asyncgens = weakref.WeakSet()
+
```
Please remove this empty line; it wasn't there to visually group all stuff related to asyncgens.

aeros replied:
Oh okay, I just thought it would be useful to group the shutdown flags visually. I'll remove it.

1st1 replied:
Yeah, sorry for the nitpicking :)

aeros replied:
@1st1:
> Yeah, sorry for the nitpicking :)

No problem, I don't mind nitpicking. I can definitely understand why you're tight on the formatting changes, it helps to keep the code more organized. (:

1st1 commented on Lib/asyncio/base_events.py:
```diff
@@ -543,6 +550,16 @@ def _asyncgen_firstiter_hook(self, agen):
                 })
 
+    def shutdown_default_executor(self):
+        """Shutdown the default executor, but wait for the threads
```
You need to fit the first sentence into one line.

1st1 replied:
To avoid code duplication, you can add a `def _shutdown_default_executor(self, wait):` and call it from `loop.close()` and `loop. shutdown_default_executor`

aeros replied:
@1st1:
> You need to fit the first sentence into one line.

I think that would exceed the recommended 80 char limit per line, would you like me to fit the max words within the limit on each line or reduce the docstring sentence? I can make it a bit shorter:

> """Shutdown the default executor, but wait for the threadpool to finish."""

1st1 replied:
> """Shutdown the default executor, but wait for the threadpool to finish."""

That works.  In general, we try to use this rule: keep the first sentence short and on one line.  Add another paragraph if needed to provide more details.

A lot of tools that work with docstrings/etc like when the first line is complete.

aeros replied:
Ah, good to know. Thanks.

1st1 commented on Lib/asyncio/events.py:
```diff
@@ -249,6 +249,12 @@ def close(self):
+    def shutdown_default_executor(self):
+        """Shutdown the default executor, but wait for the threads
+        in the threadpool to finish joining.
+        """
```
ditto

aeros replied:
See above comment

aeros committed:
Remove empty line

aeros committed:
Shorten `loop.shutdown_default_executor()` docstring

aeros committed:
Add _shutdown_default_executor() to reduce duplicate code

asvetlov reviewed (approved):
LGTM
Should the following PR make `shutdown_default_executor()` an async method?

aeros committed:
Update documentation

aeros committed:
Convert `loop.shutdown_default_executor` into a coroutine

aeros committed:
Add internal method `_do_shutdown()` and fix flag in `shutdown_default_executor()

aeros committed:
Update docstrings to reflect changes to `loop.shutdown_default_executor`

aeros committed:
Misc fixes for `loop._do_shutdown()`

aeros committed:
Fix whitespace in `asyncio-eventloop.rst`

aeros committed:
Update `asyncio.run()` for `loop.shutdown_default_executor()` coroutine improvement

aeros committed:
Fix `loop.shutdown_default_executor()` and whitespace

aeros committed:
Delete accidentally committed temp file

aeros commented on Lib/asyncio/runners.py:
```diff
@@ -45,6 +45,7 @@ def run(main, *, debug=False):
         try:
             _cancel_all_tasks(loop)
             loop.run_until_complete(loop.shutdown_asyncgens())
+            loop.run_until_complete(loop.shutdown_default_executor())
```
Since this is now a coroutine, it should be called with ``loop.run_until_complete()`` instead of by itself as it was previously.

asvetlov replied:
That's fine

aeros commented on Lib/asyncio/base_events.py:
```diff
@@ -543,6 +549,26 @@ def _asyncgen_firstiter_hook(self, agen):
+    async def shutdown_default_executor(self):
+        """Schedule the shutdown of the default executor."""
+        self._executor_shutdown_called = True
+        if self._default_executor is None:
```
This check is needed, otherwise numerous exceptions occur from ``test_asyncio``. There's nothing to shutdown if the default executor isn't set.

aeros commented on Lib/asyncio/events.py:
```diff
@@ -249,6 +249,10 @@ def close(self):
         raise NotImplementedError
 
+    async def shutdown_default_executor(self):
+        """Schedule the shutdown of the default executor."""
```
I changed the docstrings to reflect that this method is now a coroutine. There might be a better way to word this, but I had trouble thinking of something more descriptive that didn't go beyond a single line. Let me know if the previous one is preferable over this one.

asvetlov replied:
The docstring is fine

aeros commented on Doc/library/asyncio-task.rst:
```diff
@@ -229,6 +229,8 @@ Running an asyncio Program
        **Important:** this function has been added to asyncio in
        Python 3.7 on a :term:`provisional basis <provisional api>`.
```
@1st1 Is ``asyncio.run()`` still going to be considered provisional in 3.9?

asvetlov replied:
Surely no. 
It should be fixed in a separate PR.

aeros replied:
I'll leave this unresolved so it's easier to find later.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

asvetlov commented on Lib/asyncio/base_events.py:
```diff
@@ -406,6 +406,8 @@ def __init__(self):
         # Set to True when `loop.shutdown_asyncgens` is called.
         self._asyncgens_shutdown_called = False
+        # Set to True when `loop.shutdown_default_executor` is called.
+        self._executor_shutdown_called = False
```
You don't need the flag.
Assign `self._default_executor = None` after shutdown to make `shutdown_default_executor()` idempotent, that's it.

aeros replied:
Resolving this comment based on my other comment: https://github.com/python/cpython/pull/15735#discussion_r323894385. Feel free to reopen it if you think this still needs to be addressed.

asvetlov commented on Lib/asyncio/base_events.py:
```diff
@@ -768,6 +795,8 @@ def run_in_executor(self, executor, func, *args):
         if executor is None:
             executor = self._default_executor
+            # Only check when the default executor is being used
+            self._check_default_executor()
```
I doubt if we need the check.
`shutdown_default_executor()` is called at the very end of event loop lifecycle, all tasks are stopped and killed at this point.
There is no code that can call `run_in_executor()` at the moment.

aeros replied:
@asvetlov 
> I doubt if we need the check.

> You don't need the flag.

Okay, I'll try removing both of them, running the tests again, and observing what happens. I think the main thing that might have been causing the earlier issues was the lack of a ``if executor is None`` check for ``shutdown_default_executor()``, which was addressed.

aeros replied:
@asvetlov 
>  shutdown_default_executor() is called at the very end of event loop lifecycle, all tasks are stopped and killed at this point.
There is no code that can call run_in_executor() at the moment.

I think the concern earlier from @1st1 was users directly calling ``shutdown_default_executor()`` and then trying to call ``run_in_executor()`` at a later point:

>  Just like with shutdown_asyncgens() -- it would be invalid to call loop.run_in_executer() after loop.shutdown_threadpool() is called.

(this was before the rename was suggested)

Does this no longer apply?

asvetlov replied:
Well, makes sense. Sorry

aeros replied:
@asvetlov 
> Well, makes sense. Sorry

No problem. If we ever phase out the lower level API from the public in the distant future and make it internal, it might be worth removing the flag and check, but as long as ``run_in_executor()`` is part of the public API, it probably needs to remain.

asvetlov reviewed (approved):
LGTM

1st1 commented on Doc/library/asyncio-eventloop.rst:
```diff
@@ -167,6 +167,18 @@ Running and stopping the loop
 
+.. coroutinemethod:: loop.shutdown_default_executor()
+
+   Schedule the closure of the default executor, but wait for the
```
```suggestion
   Schedule the closure of the default executor and wait for it to join all threads
```

1st1 reviewed (approved):
Please fix the last remaining nit I pointed out.  Thanks!  Looks great

aeros committed:
Adjust wording in doc for loop.shutdown_default_executor()

Co-Authored-By: Yury Selivanov <yury@edgedb.com>

aeros committed:
Adjust wording in Misc/NEWS entry

Co-Authored-By: Yury Selivanov <yury@edgedb.com>

aeros committed:
Merge branch 'bpo-34037' of https://github.com/aeros167/cpython into bpo-34037

aeros committed:
Fix doc for loop.shutdown_default_executor()

After applying the suggestion from @1st1, this adjusts the surrounding wording and flow of the paragraph.

aeros committed:
Add "Patch by" to news entry
